### PR TITLE
Reject queries with overlength time-range

### LINKED ISSUE #349: QoD caused by start=<huge-negative-value>&end=<positive-value> (huge start-end range)
Follow up of: https://github.com/weaveworks/cortex/issues/327

My QoD is still valid. I dived into the issue and it seems that the confusion here is because Prometheus wrongly parses my mistaken `start=1489025420000` timestamp. From query perspective it is not really start > time problem, but actually underlying int64 overflow I fixed here:  https://github.com/prometheus/prometheus/pull/2501 caused start to be huge negative value.

This means that QoD is still possible and is caused not by `start > end`, but actually huge range (e.g `start=MinInt64&end=MaxInt64`)

@tomwilkie do you have any idea how to handle this properly? IMO Some limiter or further validation is needed.

### PR DESCRIPTION
**Pull request by meccanorak**
Should resolve #349

~~Limit is configured via new command line option -max-query-range-days=~~
~~If option omitted, limit defaults to 400 days (just over 13 months)~~
Limit is configured via new command line option so to set limit
to 30 days would be -querier.max-time-range=720h
(Unfortunately that time.Duration parsing doesn't accept day/week/month/year as units)
If option omitted, limit defaults to 400 days (just over 13 months)
For no limit on time-range, configure to 0

All comment welcome, but particularly ...

- Ought name of command line option be of form -abcxyz.max-query-range-days rather than simply -max-query-range-days ?  In which case, what substitutes for abcxyz?
- Is StoreConfig a good choice to hold new option value ... or would it make more sense to store it in some other Config object?
- As this changeset stands, new testing is only against v8schema .... would testing against all schemas be preferable?

### CODE REVIEW

**Comment by aaron7** [comment]
This is great, thanks!

> Is StoreConfig a good choice to hold new option value ... or would it make more sense to store it in some other Config object?

Perhaps we could create a QuerierConfig? This would include arguments related to querying, such as max range, timeout etc.

> Ought name of command line option be of form -abcxyz.max-query-range-days rather than simply -max-query-range-days ? In which case, what substitutes for abcxyz?

If we created the querier config, it would be `-querier.max-range-days` - which is clear.

> As this changeset stands, new testing is only against v8schema .... would testing against all schemas be preferable?

Yes, testing against all schema's is preferable (many other tests do this, so you should be able to use similar code). The old schema are still used when querying old data and it also reminds us to test this against any new schema we add (we would probably miss it if this test hardcoded v8Schema)

File: pkg/chunk/chunk_store.go [hunk 1]
```diff
@@ -154,6 +160,14 @@ func (c *Store) Get(ctx context.Context, from, through model.Time, allMatchers .
 	if through < from {
 		return nil, fmt.Errorf("invalid query, through < from (%d < %d)", through, from)
 	}
+	if c.cfg.MaxQueryRangeDays == 0 {
+		// Specified as unlimited
+	} else {
```
Comment (aaron7) [inline pkg/chunk/chunk_store.go]: you can replace these 3 lines with just:

`if c.cfg.MaxQueryRangeDays != 0 {`

File: pkg/chunk/chunk_store.go [hunk 2]
```diff
@@ -44,13 +45,18 @@ func init() {
...
+	MaxQueryRangeDays uint
+
 	// For injecting different schemas in tests.
 	schemaFactory func(cfg SchemaConfig) Schema
 }
 
 // RegisterFlags adds the flags required to config this to the given FlagSet
 func (cfg *StoreConfig) RegisterFlags(f *flag.FlagSet) {
 	cfg.CacheConfig.RegisterFlags(f)
+	f.UintVar(&cfg.MaxQueryRangeDays, "max-query-range-days",
```
Comment (aaron7) [inline pkg/chunk/chunk_store.go]: Please use `f.DurationVar` which parses args like `100d` into `time.Duration` for you.

File: pkg/chunk/chunk_store.go [hunk 3]
```diff
@@ -154,6 +160,14 @@ func (c *Store) Get(ctx context.Context, from, through model.Time, allMatchers .
 	if through < from {
 		return nil, fmt.Errorf("invalid query, through < from (%d < %d)", through, from)
 	}
+	if c.cfg.MaxQueryRangeDays == 0 {
+		// Specified as unlimited
+	} else {
+		if through.After(from.Add(time.Hour * 24 * time.Duration(c.cfg.MaxQueryRangeDays))) {
```
Comment (aaron7) [inline pkg/chunk/chunk_store.go]: you will just need to use `c.cfg.MaxQueryRangeDays` as the value after switching to `f.DurationVar`.

File: pkg/chunk/chunk_store.go (hunk 2 above)
Comment (meccanorak) [inline pkg/chunk/chunk_store.go]: Thanks ... simpler and more conventional
+ I'll do something about the    days    suffix which no longer fits

File: pkg/chunk/chunk_store.go (hunk 1 above)
Comment (meccanorak) [inline pkg/chunk/chunk_store.go]: Yes shorter form is same logic ... but I regard longer form as more expressive to those who come after

File: pkg/chunk/chunk_store.go (hunk 3 above)
Comment (meccanorak) [inline pkg/chunk/chunk_store.go]: Absolutely

**Comment by meccanorak** [comment]
I've revised changeset to take on board review suggestions from @aaron7 
Not sure all of them are inline within expectations
One slight downside is that 'flag.FlagSet.DurationVar' parser only accepts units of hour or smaller.
It can't handle day/week/month/year
So 13ish months has to specified as  '-querier.max-time-range=9600h'

**Comment by bboreham** [comment]
Sorry this has languished so long.

I think this proposed change is in the wrong place. The limit on length of date range should be done in the Prometheus code that first unpacks the query request.  If Prometheus maintainers reject that PR then we can think again.

It would be appropriate to have a limit in the chunk store on things that are uniquely chunk store concerns, e.g. how many buckets the query touches.

**Pull request closed without merging**